src/java.base/share/classes/jdk/internal/ref/CleanerImpl.java

```diff
@@ -231,4 +223,138 @@ protected void performCleanup() {
+        public synchronized void insert(PhantomCleanable<?> phc) {
+            if (head.size == NODE_CAPACITY) {
+                // Head node is full, insert new one.
+                // If possible, pick a pre-allocated node from cache.
+                Node newHead;
+                if (cache != null) {
+                    newHead = cache;
+                    cache = null;
+                } else {
+                    newHead = new Node();
+                }
+                newHead.next = head;
+                head.prev = newHead;
+                head = newHead;
+            }
+            assert head.size < NODE_CAPACITY;
+
+            // Put the incoming object in head node and record indexes.
+            phc.node = head;
+            phc.index = head.size;
+            head.arr[head.size] = phc;
+            head.size++;
+        }
+
+        /**
+         * Remove this PhantomCleanable from the list.
+         *
+         * @return true if Cleanable was removed or false if not because
+         * it had already been removed before
+         */
+        public synchronized boolean remove(PhantomCleanable<?> phc) {
+            if (phc.node == null) {
+                // Not in the list.
+                return false;
+            }
+            assert phc.node.arr[phc.index] == phc;
+            assert head.size > 0;
+
+            // Replace with another element from the head node, as long
+            // as it is not the same element. This keeps all non-head
+            // nodes at full capacity.
+            if (head != phc.node || (phc.index != head.size - 1)) {
+                PhantomCleanable<?> mover = head.arr[head.size - 1];
+                mover.node = phc.node;
+                mover.index = phc.index;
+                phc.node.arr[phc.index] = mover;
+            }
+
+            // Now we can unlink the removed element.
+            phc.node = null;
+
+            // Remove the last element from the head node.
+            head.arr[head.size - 1] = null;
+            head.size--;
+
+            // If head node becomes empty after this, and there are
+            // nodes that follow it, replace the head node with another
+            // full one. If needed, stash the now free node in cache.
+            if (head.size == 0 && head.next != null) {
+                Node newHead = head.next;
+                newHead.prev = null;
+                if (cache == null) {
+                    cache = head;
+                    cache.prev = null;
+                    cache.next = null;
+                }
+                head = newHead;
+            }
+
+            return true;
+        }
+
+        /**
+         * Segment node.
+         */
+        static class Node {
+            // Array of tracked cleanables, and the amount of elements in it.
+            final PhantomCleanable<?>[] arr = new PhantomCleanable<?>[NODE_CAPACITY];
+            int size;
+
+            // Linked list structure.
+            Node prev;
```

Comment:
The prev link seems only ever written but never read.  This can be a singly-linked list, I believe.